vyper/builtin_functions/functions.py

```diff
@@ -1256,40 +1335,56 @@ def build_IR(self, expr, args, kwargs, contact):
 class RawLog(_BuiltinFunction):
 
     _id = "raw_log"
-    _inputs = [("topics", "*"), ("data", ("bytes32", "Bytes"))]
+    _inputs = [
+        ("topics", DynamicArrayDefinition(Bytes32Definition(), 4)),
+        ("data", (Bytes32Definition(), BytesArrayPrimitive())),
+    ]
 
     def fetch_call_return(self, node):
+        self.infer_arg_types(node)
+
+    def infer_arg_types(self, node):
         validate_call_args(node, 2)
         if not isinstance(node.args[0], vy_ast.List) or len(node.args[0].elements) > 4:
             raise InvalidType("Expecting a list of 0-4 topics as first argument", node.args[0])
+
         if node.args[0].elements:
-            validate_expected_type(
-                node.args[0], ArrayDefinition(Bytes32Definition(), len(node.args[0].elements))
+            validate_expected_type(node.args[0], self._inputs[0][1])
+
+        data_type = get_possible_types_from_node(node.args[1]).pop()
+
+        if not isinstance(data_type, BytesAbstractType):
```

Comment:
looks like duplication of `validate_expected_type` logic. why not delegate down to `validate_expected_type`?